Resolve "Text data type does not delete as expected"

jbdyn opened the pull request:
Essentially, I adjust the conditions for range lengths such that they align with the signature of the `remove_range` method of `TextRef` in Yrs.

Additionally, I review convenience methods for text manipulation `replace_range`, `remove_range`, `insert`, and `clear`, inspired by the [blog post of Darren Burns](https://textual.textualize.io/blog/2023/09/18/things-i-learned-while-building-textuals-textarea/#edits-are-replacements)

davidbrochart commented:
Thanks for opening a PR.
What do you mean by "Text data type does not delete as expected"? Can you add a test showing that it solves an issue?

jbdyn commented:
Sure!

I added an [example](https://github.com/jupyter-server/pycrdt/issues/67#issue-2089807201) in the corresponding issue and I am about to write tests atm.

Do you agree with the extra methods I added?

davidbrochart commented:
I didn't see the issue, sorry.
I'll have a look next week.

davidbrochart commented on python/pycrdt/text.py:
```diff
@@ -103,19 +103,30 @@ def __setitem__(self, key: int | slice, value: str) -> None:
+        del self[start:stop]
+        self[start:start] = value
+
+    def remove_range(self, start: int, stop: int) -> None:
```
This method doesn't exist on a builtin Python `list`, and it does the same as:
```py
del text[start:stop]
```
Do you think it is worth it?

jbdyn replied:
I would think so. I like to have `insert` and `clear` as verbose option since they carry the reading flow better than brackets and colons, in my opinion.
Another benefit would be that a dedicated method pops up in auto-completion, which might come in handy for new users so they know about this without the need to read a documentation.

davidbrochart replied:
But `insert` and `clear` are methods of a `list`, and `remove_range` is not.
I'd like to keep the API as small as possible for now.

jbdyn replied:
I'm totally fine with that.

I reverted the changes.

davidbrochart commented on python/pycrdt/text.py:
```diff
@@ -103,19 +103,30 @@ def __setitem__(self, key: int | slice, value: str) -> None:
             else:
                 raise RuntimeError(f"Index not supported: {key}")
 
+    def replace_range(self, value: str, start: int, stop: int) -> None:
```
This method doesn't exist on a builtin `list`, and it does the same as:
```py
text[start:stop] = value
```
Do you think it is worth it?

jbdyn replied:
Please the [above](https://github.com/jupyter-server/pycrdt/pull/68#discussion_r1461774125)

davidbrochart replied:
Same comment as for `remove_range`.

jbdyn replied:
Reverted.

davidbrochart commented on python/pycrdt/text.py:
```diff
@@ -72,22 +72,22 @@ def _check_slice(self, key: slice) -> tuple[int, int]:
-                    self.integrated.remove_range(txn._txn, start, stop)
+                start, length = self._check_slice(key)
+                if length > 0:
+                    self.integrated.remove_range(txn._txn, start, length)
```
I'd rather have `_check_slice` still return `start, stop` and compute the `length` here, as we might support `step` in the future.

davidbrochart reviewed (commented):
Thanks @jbdyn, I left a few comments.

jbdyn commented on python/pycrdt/text.py:
```diff
@@ -103,19 +103,30 @@ def __setitem__(self, key: int | slice, value: str) -> None:
+        """Remove the range of characters in the interval ['start', 'stop')."""
+        del self[start:stop]
+
+    def insert(self, value: str, index: int) -> None:
```
For me, reading this like "insert 'value' at index 'index'" feels just a bit more natural than "insert at index 'index' 'value'". I also wanted it to be similar to my suggested `replace_range`, which would have the same signature as [`replace` from Textual's `TextArea` widget](https://github.com/Textualize/textual/blob/9f43d300052f3a2ef8490a522e41463897477136/src/textual/widgets/_text_area.py#L1627). I thought: Since `insert` and `clear` are somewhat verbose convenience functions, I make them a bit more convenient for me.
However, I realized just now that the current implementation follows the signature also used in `list.insert` as well as in the underlying Yrs `TextRef` `insert` method.

Swapping arguments here would be inconsistent and thus might be more confusing than helpful.

What do you think?

davidbrochart reviewed (approved):
Thanks a lot @jbdyn !
